substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVFileTableRecord.java

```diff
@@ -0,0 +1,140 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.objectfile.pecoff.cv;
+
+import com.oracle.objectfile.debugentry.FileEntry;
+import org.graalvm.compiler.debug.GraalError;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+final class CVFileTableRecord extends CVSymbolRecord {
+
+    private static final byte CHECKSUM_NONE = 0x00;
+    private static final byte CHECKSUM_MD5 = 0x01;
+    private static final byte CB_VALUE = 0x10;
+
+    private static final int FILE_RECORD_LENGTH = 24;
+
+    private static final int CHECKSUM_LENGTH = 16;
+    private static final byte[] EMPTY_CHECKSUM = new byte[CHECKSUM_LENGTH];
+
+    private static final int FILE_TABLE_INITIAL_SIZE = 200;
+
+    private final CVSymbolSectionImpl.CVStringTable strings;
+
+    private int currentOffset = 0;
+    private Map<FileEntry, Integer> fileEntryToOffsetMap = new LinkedHashMap<>(FILE_TABLE_INITIAL_SIZE);
+
+    CVFileTableRecord(CVDebugInfo cvDebugInfo, CVSymbolSectionImpl.CVStringTable strings) {
+        super(cvDebugInfo, CVDebugConstants.DEBUG_S_FILECHKSMS);
+        this.strings = strings;
+    }
+
+    int addFile(FileEntry entry) {
+        if (fileEntryToOffsetMap.containsKey(entry)) {
+            return fileEntryToOffsetMap.get(entry);
+        } else {
+            fileEntryToOffsetMap.put(entry, currentOffset);
+            /* Create required stringtable entry. */
+            strings.add(fixPath(entry));
+            currentOffset += FILE_RECORD_LENGTH;
+            return currentOffset - FILE_RECORD_LENGTH;
+        }
+    }
+
+    @Override
+    public int computeSize(int initialPos) {
+        /*
+         * As part of computing the size, add all fileEntries ignoring duplicates. Most fileEntries
+         * are created while building the symbol section.
+         */
+        for (FileEntry entry : cvDebugInfo.getFiles()) {
+            addFile(entry);
```

Comment:
As far as I can see, if there is a new file here, it means that there is no line entry that refers to it, so why add it? How will it be used?

Reply:
This is partially defensive programming, and partially to account for files that define types without code, which will be used when the type section PR comes along.